presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java

```diff
@@ -25,24 +26,36 @@
 public class KerberosHadoopAuthentication
         implements HadoopAuthentication
 {
-    static {
+    private final KerberosAuthentication kerberosAuthentication;
+
+    public static KerberosHadoopAuthentication createKerberosHadoopAuthentication(KerberosAuthentication kerberosAuthentication, HdfsConfiguration hdfsConfiguration)
+    {
+        KerberosHadoopAuthentication kerberosHadoopAuthentication = new KerberosHadoopAuthentication(kerberosAuthentication);
+
+        Configuration configuration = requireNonNull(hdfsConfiguration, "hdfsConfiguration is null").getConfiguration();
+        initializeConfigs(configuration);
+        return kerberosHadoopAuthentication;
+    }
+
+    private KerberosHadoopAuthentication(KerberosAuthentication kerberosAuthentication)
+    {
+        this.kerberosAuthentication = requireNonNull(kerberosAuthentication, "kerberosAuthentication is null");
+    }
+
+    private static void initializeConfigs(Configuration configuration)
+    {
         // In order to enable KERBEROS authentication method for HDFS
         // UserGroupInformation.authenticationMethod static field must be set to KERBEROS
         // It is further used in many places in DfsClient
-        Configuration configuration = new Configuration(false);
         configuration.set("hadoop.security.authentication", "kerberos");
-        UserGroupInformation.setConfiguration(configuration);
 
-        // KerberosName#rules static field must be initialized
-        // It is used in KerberosName#getShortName which is used in User constructor invoked by UserGroupInformation#getUGIFromSubject
-        KerberosName.setRules("DEFAULT");
-    }
+        String authToLocalRules = configuration.get("hadoop.security.auth_to_local");
```

Comment:
why not use `org.apache.hadoop.security.HadoopKerberosName#setConfiguration` instead?

Reply:
When i was implementing this, i wanted to get to the bottom of that nasty code, and leave it as obvious as possible. That's why i didn't use `HadoopKerberosName#setConfiguration`. Please don't change it.

Reply:
This method is actually being called from the inside of the `UserGroupInformation.setConfiguration`. By setting the rules to `DEFAULT` explicitly i wanted to explicitly depict the statistic components, that are being set. You can just remove the `KerberosName.setRules("DEFAULT")` all together.